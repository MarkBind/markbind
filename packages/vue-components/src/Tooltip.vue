<template>
  <span
    :class="trigger === 'click' ? 'trigger-click' : 'trigger'"
    data-mb-component-type="tooltip"
    tabindex="0"
  >
    <portal v-if="targetEl.id" :to="'tooltip:' + targetEl.id">
      <slot name="content"></slot>
    </portal>

    <v-tooltip
      v-if="isMounted"
      :auto-hide="!isInput"
      :triggers="triggers"
      :popper-triggers="triggers"
      :hide-triggers="triggers"
      :placement="placement"
      :delay="0"
      shift-cross-axis
    >
      <template #popper>
        <slot name="content"></slot>
      </template>

      <span v-if="!isInput" @click.stop>
        <slot></slot>
      </span>
      <slot v-else></slot>
    </v-tooltip>
  </span>
</template>

<script>
// eslint-disable-next-line import/no-extraneous-dependencies
import { Portal } from 'portal-vue';

export default {
  name: 'Tooltip',
  components: {
    Portal,
  },
  props: {
    trigger: {
      type: String,
      default: 'hover focus',
    },
    placement: {
      type: String,
      default: 'top',
    },
  },
  data() {
    return {
      targetEl: {},
      isInput: false,
      isMounted: false,
    };
  },
  computed: {
    triggers() {
      return this.trigger.split(' ');
    },
  },
  mounted() {
    this.targetEl = this.$el;
    // <input> tags need to be handled separately as they need to retain focus on inputs
    this.isInput = this.$slots.default && this.$slots.default.some(node => node.tag === 'input');
    this.isMounted = true;
  },
};
</script>

<style>
    /* stylelint-disable selector-class-pattern */
    .v-popper--theme-tooltip .v-popper__inner {
        /* following bootstrap */
        background: rgb(0 0 0 / 90%);
        padding: 4px 8px;
        font-size: 0.875rem;
        max-width: 200px;
        text-align: center;
    }
    /* stylelint-enable selector-class-pattern */
</style>
