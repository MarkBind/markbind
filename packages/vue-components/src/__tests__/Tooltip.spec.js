import { mount } from '@vue/test-utils';
import Tooltip from '../Tooltip.vue';

const DEFAULT_TOOLTIP_TRIGGER = 'click to show tooltip';
const DEFAULT_TOOLTIP_CONTENT = 'Lorem ipsum dolor sit amet';

describe('Tooltip', () => {
  test('should not be shown until triggered', async () => {
    const wrapper = mount(Tooltip, {
      propsData: {
        trigger: 'click',
        content: DEFAULT_TOOLTIP_CONTENT,
      },
      slots: {
        default: DEFAULT_TOOLTIP_TRIGGER,
      },
      stubs: ['v-tooltip'],
    });

    // don't click on trigger
    expect(wrapper.element).toMatchSnapshot();
  });

  test('should show contents correctly when triggered', async () => {
    const wrapper = mount(Tooltip, {
      propsData: {
        trigger: 'click',
        content: DEFAULT_TOOLTIP_CONTENT,
      },
      slots: {
        default: DEFAULT_TOOLTIP_TRIGGER,
      },
      stubs: ['v-tooltip'],
    });

    // click on trigger
    await wrapper.find('span.trigger-click').trigger('click');

    expect(wrapper.element).toMatchSnapshot();
  });

  test('should respect the placement attribute', async () => {
    const wrapper = mount(Tooltip, {
      propsData: {
        trigger: 'click',
        content: DEFAULT_TOOLTIP_CONTENT,
        placement: 'right',
      },
      slots: {
        default: DEFAULT_TOOLTIP_TRIGGER,
      },
      stubs: ['v-tooltip'],
    });

    // click on trigger
    await wrapper.find('span.trigger-click').trigger('click');

    expect(wrapper.element).toMatchSnapshot();
  });
});
