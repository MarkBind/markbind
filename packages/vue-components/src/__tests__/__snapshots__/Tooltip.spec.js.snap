// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should not be shown until triggered 1`] = `
<span
  class="trigger-click"
  content="Lorem ipsum dolor sit amet"
  data-mb-component-type="tooltip"
  tabindex="0"
>
  <!---->
  <!---->
</span>
`;

exports[`Tooltip should respect the placement attribute 1`] = `
<span
  class="trigger-click"
  content="Lorem ipsum dolor sit amet"
  data-mb-component-type="tooltip"
  tabindex="0"
>
  <!---->
  <v-tooltip-stub
    auto-hide="true"
    delay="0"
    hide-triggers="click"
    placement="right"
    popper-triggers="click"
    shift-cross-axis=""
    triggers="click"
  >
    <span>
      click to show tooltip
    </span>
  </v-tooltip-stub>
</span>
`;

exports[`Tooltip should show contents correctly when triggered 1`] = `
<span
  class="trigger-click"
  content="Lorem ipsum dolor sit amet"
  data-mb-component-type="tooltip"
  tabindex="0"
>
  <!---->
  <v-tooltip-stub
    auto-hide="true"
    delay="0"
    hide-triggers="click"
    placement="top"
    popper-triggers="click"
    shift-cross-axis=""
    triggers="click"
  >
    <span>
      click to show tooltip
    </span>
  </v-tooltip-stub>
</span>
`;
